bitcoin/src/blockdata/transaction.rs

```diff
@@ -2444,6 +2444,26 @@ mod tests {
         let seq = Sequence::from_seconds_floor(1000);
         println!("{:?}", seq)
     }
+
+    #[test]
+    fn debug_txid_has_no_prefix() {
+        let outpoint = OutPoint::default();
+
+        assert_eq!("OutPoint { txid: 0000000000000000000000000000000000000000000000000000000000000000, vout: 4294967295 }", format!("{:?}", &outpoint));
+        assert_eq!(
+            "0000000000000000000000000000000000000000000000000000000000000000:4294967295",
+            format!("{}", &outpoint)
+        );
+        assert_eq!(
+            "OutPoint {
+    txid: 0000000000000000000000000000000000000000000000000000000000000000,
+    vout: 4294967295,
+}",
+            format!("{:#?}", &outpoint)
+        );
+        assert_eq!(
+            "0x0000000000000000000000000000000000000000000000000000000000000000", format!("{:#}", &outpoint.txid));
+    }
```

Comment:
To test the `0x` prefix I think we should have a unit test in `hashes` because that is where the code is. To test the pretty printing of `Outpoint` (which IMO is a valid and valuable test) the test is good here in `transaction`.

I would have expected something like this in `hashes/src/lib.rs`
```rust

    #[test]
    fn display() {
        let want = "0000000000000000000000000000000000000000000000000000000000000000";
        let got = format!("{}", sha256d::Hash::all_zeros());
        assert_eq!(got, want);
    }

    #[test]
    fn debug() {
        let want = "0000000000000000000000000000000000000000000000000000000000000000";
        let got = format!("{:?}", sha256d::Hash::all_zeros());
        assert_eq!(got, want);
    }

    #[test]
    fn alternate_hex() {
        let want = "0x0000000000000000000000000000000000000000000000000000000000000000";
        let got = format!("{:#x}", sha256d::Hash::all_zeros());
        assert_eq!(got, want);
    }

    #[test]
    fn alternate_pretty() {
        let want = "0000000000000000000000000000000000000000000000000000000000000000";
        let got = format!("{:#?}", sha256d::Hash::all_zeros());
        assert_eq!(got, want);
    }
```

And in `transaction.rs`:

```rust
    #[test]
    fn outpoint_format() {
        let outpoint = OutPoint::default();

        let display = "0000000000000000000000000000000000000000000000000000000000000000:4294967295";
        let pretty = "OutPoint {
    txid: 0000000000000000000000000000000000000000000000000000000000000000,
    vout: 4294967295,
}";
        let got = format!("{:?}", &outpoint);
        assert_eq!(got, pretty);

        let got = format!("{}", &outpoint);
        assert_eq!(got, display);

        let got = format!("{:#}", &outpoint);
        assert_eq!(got, display);

        let got = format!("{:?}", &outpoint);
        assert_eq!(got, pretty);

        let got = format!("{:#?}", &outpoint);
        assert_eq!(got, pretty);
    }
```

Reply:
Thanks for reviewing my work @apoelstra @tcharding .

For the tests suggested in file `src/hashes/lib.rs` , they are already included in the `hashes/src/utils.rs` from line https://github.com/rust-bitcoin/rust-bitcoin/blob/67c80f197fac24b13a2b65f85672038cc1e7df2d/hashes/src/util.rs#L421

This is because they implement the macro rule `hash_newtype` at  https://github.com/rust-bitcoin/rust-bitcoin/blob/67c80f197fac24b13a2b65f85672038cc1e7df2d/hashes/src/util.rs#L180 which call the other macro rule `hex_fmt_impl` at line https://github.com/rust-bitcoin/rust-bitcoin/blob/67c80f197fac24b13a2b65f85672038cc1e7df2d/hashes/src/util.rs#L5C14-L5C26 with modified changes at line 43.

Is my reasoning behind those tests correct?

Reply:
https://github.com/rust-bitcoin/rust-bitcoin/commit/3c01be022f5c1af818a5e90391733a15d28a348e improves the tests for `transaction.rs` from the recommendation you gave